In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns


In [3]:
import sys
print(sys.executable)

c:\Python313\python.exe


In [4]:
df=pd.read_csv('../data/raw/fake_job_postings.csv')
df.head()

,job_id,title,location,department,salary_range,company_profile,description,requirements,benefits,telecommuting,has_company_logo,has_questions,employment_type,required_experience,required_education,industry,function,fraudulent
0,1,Marketing Intern,"US, NY, New York",Marketing,NaN,"We're Food52, and we've created a groundbreaki...","Food52, a fast-growing, James Beard Award-winn...",Experience with content management systems a m...,NaN,0,1,0,Other,Internship,NaN,NaN,Marketing,0
1,2,Customer Service - Cloud Video Production,"NZ, , Auckland",Success,NaN,"90 Seconds, the worlds Cloud Video Production ...",Organised - Focused - Vibrant - Awesome!Do you...,What we expect from you:Your key responsibilit...,What you will get from usThrough being part of...,0,1,0,Full-time,Not Applicable,NaN,Marketing and Advertising,Customer Service,0
2,3,Commissioning Machinery Assistant (CMA),"US, IA, Wever",NaN,NaN,Valor Services provides Workforce Solutions th...,"Our client, located in Houston, is actively se...",Implement pre-commissioning and commissioning ...,NaN,0,1,0,NaN,NaN,NaN,NaN,NaN,0
3,4,Account Executive - Washington DC,"US, DC, Washington",Sales,NaN,Our passion for improving quality of life thro...,THE COMPANY: ESRI – Environmental Systems Rese...,"EDUCATION: Bachelor’s or Master’s in GIS, busi...",Our culture is anything but corporate—we have ...,0,1,0,Full-time,Mid-Senior level,Bachelor's Degree,Computer Software,Sales,0
4,5,Bill Review Manager,"US, FL, Fort Worth",NaN,NaN,SpotSource Solutions LLC is a Global Human Cap...,JOB TITLE: Itemization Review ManagerLOCATION:...,QUALIFICATIONS:RN license in the State of Texa...,Full Benefits Offered,0,1,1,Full-time,Mid-Senior level,Bachelor's Degree,Hospital & Health Care,Health Care Provider,0


In [5]:
columns=df.columns
columns

Index(['job_id', 'title', 'location', 'department', 'salary_range',
       'company_profile', 'description', 'requirements', 'benefits',
       'telecommuting', 'has_company_logo', 'has_questions', 'employment_type',
       'required_experience', 'required_education', 'industry', 'function',
       'fraudulent'],
      dtype='object')

In [6]:
df.shape

(17880, 18)

In [7]:
df.describe()

,job_id,telecommuting,has_company_logo,has_questions,fraudulent
count,17880.000000,17880.000000,17880.000000,17880.000000,17880.000000
mean,8940.500000,0.042897,0.795302,0.491723,0.048434
std,5161.655742,0.202631,0.403492,0.499945,0.214688
min,1.000000,0.000000,0.000000,0.000000,0.000000
25%,4470.750000,0.000000,1.000000,0.000000,0.000000
50%,8940.500000,0.000000,1.000000,0.000000,0.000000
75%,13410.250000,0.000000,1.000000,1.000000,0.000000
max,17880.000000,1.000000,1.000000,1.000000,1.000000


In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17880 entries, 0 to 17879
Data columns (total 18 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   job_id               17880 non-null  int64 
 1   title                17880 non-null  object
 2   location             17534 non-null  object
 3   department           6333 non-null   object
 4   salary_range         2868 non-null   object
 5   company_profile      14572 non-null  object
 6   description          17879 non-null  object
 7   requirements         15184 non-null  object
 8   benefits             10668 non-null  object
 9   telecommuting        17880 non-null  int64 
 10  has_company_logo     17880 non-null  int64 
 11  has_questions        17880 non-null  int64 
 12  employment_type      14409 non-null  object
 13  required_experience  10830 non-null  object
 14  required_education   9775 non-null   object
 15  industry             12977 non-null  object
 16  func

In [9]:
fraudCount=(df['fraudulent']==1).sum()
fraudCount

np.int64(866)

In [10]:
missing_percentages_frauds={}
missing_percentage_Nonfrauds={}
for col in columns:
    missing_percentages_frauds[col]=((((df['fraudulent']==1) & df[col].isnull()).sum())/fraudCount)*100
    missing_percentage_Nonfrauds[col]=((((df['fraudulent']==0) & df[col].isnull()).sum())/(df.shape[0]-fraudCount))*100
print(missing_percentages_frauds)
print(missing_percentage_Nonfrauds)

{'job_id': np.float64(0.0), 'title': np.float64(0.0), 'location': np.float64(2.1939953810623556), 'department': np.float64(61.316397228637406), 'salary_range': np.float64(74.24942263279446), 'company_profile': np.float64(67.78290993071595), 'description': np.float64(0.11547344110854503), 'requirements': np.float64(17.782909930715935), 'benefits': np.float64(42.032332563510394), 'telecommuting': np.float64(0.0), 'has_company_logo': np.float64(0.0), 'has_questions': np.float64(0.0), 'employment_type': np.float64(27.829099307159353), 'required_experience': np.float64(50.23094688221709), 'required_education': np.float64(52.078521939953816), 'industry': np.float64(31.755196304849886), 'function': np.float64(38.91454965357968), 'fraudulent': np.float64(0.0)}
{'job_id': np.float64(0.0), 'title': np.float64(0.0), 'location': np.float64(1.921946632185259), 'department': np.float64(64.7466792053603), 'salary_range': np.float64(84.453979076055), 'company_profile': np.float64(15.992711884330552), 

In [11]:
compare_missing = pd.DataFrame({
    'fraud_missing_%': pd.Series(missing_percentages_frauds),
    'nonfraud_missing_%': pd.Series(missing_percentage_Nonfrauds)
})

compare_missing['difference_in_missing_percentage'] = (
    compare_missing['fraud_missing_%'] -
    compare_missing['nonfraud_missing_%']
)

compare_missing


,fraud_missing_%,nonfraud_missing_%,difference_in_missing_percentage
job_id,0.000000,0.000000,0.000000
title,0.000000,0.000000,0.000000
location,2.193995,1.921947,0.272049
department,61.316397,64.746679,-3.430282
salary_range,74.249423,84.453979,-10.204556
company_profile,67.782910,15.992712,51.790198
description,0.115473,0.000000,0.115473
requirements,17.782910,14.940637,2.842273
benefits,42.032333,40.249207,1.783126
telecommuting,0.000000,0.000000,0.000000


In [12]:
compare_missing.sort_values(by='difference_in_missing_percentage', ascending=False)

,fraud_missing_%,nonfraud_missing_%,difference_in_missing_percentage
company_profile,67.782910,15.992712,51.790198
required_experience,50.230947,38.879746,11.351201
employment_type,27.829099,18.984366,8.844733
required_education,52.078522,44.986482,7.092040
industry,31.755196,27.201128,4.554068
function,38.914550,35.958622,2.955927
requirements,17.782910,14.940637,2.842273
benefits,42.032333,40.249207,1.783126
location,2.193995,1.921947,0.272049
description,0.115473,0.000000,0.115473


In [13]:
text_cols = ['company_profile', 'description', 'requirements']

for col in text_cols:
    df[col + '_length'] = df[col].fillna('').apply(len)


In [14]:
for col in text_cols:
    print(col)
    print(
        df.groupby('fraudulent')[col + '_length'].mean()
    )
    print('-' * 30)


company_profile
fraudulent
0    640.752674
1    230.891455
Name: company_profile_length, dtype: float64
------------------------------
description
fraudulent
0    1221.219701
1    1154.834873
Name: description_length, dtype: float64
------------------------------
requirements
fraudulent
0    597.465910
1    446.049654
Name: requirements_length, dtype: float64
------------------------------


In [15]:
print(df.loc[df['fraudulent'] == 1, 'description'].sample(5, random_state=42))


17789                   customer service reps needed asap 
11628    - Maintain a high level of Clearmark solution ...
17617    Essential DutiesGreets callers and visitors, d...
17777    Hiring Part-Time Workers (CASH PAYING)You can ...
997      Responsible for assisting in the direction and...
Name: description, dtype: object


In [16]:
urgency_keywords = [
    'immediate', 'urgent', 'asap',
    'quickly', 'fast', 'fast hiring', 'limited'
]
for i in range(len(df)):
    text = df.loc[i, 'description']
    
    if pd.isna(text):
        continue  
    
    text = text.lower()
    
    for word in urgency_keywords:
        if word in text:
            df.loc[i, 'has_urgency_language'] = 1
            break



In [17]:
count_urgency_fake=((df['fraudulent']==1) & (df['has_urgency_language']==1)).sum()
count_urgency_fake

np.int64(253)

In [18]:
count_urgency_fake=((df['fraudulent']==0) & (df['has_urgency_language']==1)).sum()
count_urgency_fake

np.int64(4878)

In [ ]:
import spacy
nlp = spacy.load("en_core_web_sm")


In [21]:
df['org_count'] = 0
df['gpe_count'] = 0

for i in range(len(df)):
    text = df.loc[i, 'company_profile']
    
    if pd.isna(text):
        continue
    
    doc = nlp(text)
    
    for ent in doc.ents:
        if ent.label_ == 'ORG':
            df.loc[i, 'org_count'] += 1
        elif ent.label_ == 'GPE':
            df.loc[i, 'gpe_count'] += 1

df.groupby('fraudulent')[['org_count', 'gpe_count']].mean()


,org_count,gpe_count
fraudulent,,
0,2.242154,1.112025
1,0.812933,0.168591


In [22]:
df['lexical_diversity']=0.0
for i in range(len(df)):
    text=df.loc[i,'description']
    if pd.isna(text):
        continue
    words=text.lower().split()
    if(len(words)==0):
        continue
    else:
        unique_words=set(words)
        df.loc[i,'lexical_diversity']=len(unique_words)/len(words)
    

In [23]:
df.groupby('fraudulent')['lexical_diversity'].mean()

fraudulent
0    0.718962
1    0.742843
Name: lexical_diversity, dtype: float64

In [24]:
df['num_in_requirements']=0.0
for i in range(len(df)):
    text=df.loc[i,'requirements']
    if pd.isna(text):
        continue
    count=0
    for ch in text:
        if ch.isdigit():
            count+=1
        df.loc[i,'num_in_requirements']=count

df.groupby('fraudulent')['num_in_requirements'].mean()

KeyboardInterrupt: 

EDA Summary & Feature Decisions

- The dataset is highly imbalanced, with only 866 fraudulent postings out of 17,880 (~4.8%). This implies that accuracy alone would be misleading, since a naive model predicting all jobs as real would already achieve ~95% accuracy. Therefore, metrics such as recall and precision will be more meaningful.

- Several features contain a large number of missing values. Missing-value analysis revealed that company_profile shows a very large missingness gap (>50%) between fake and real jobs, making it a strong fraud signal. Other fields such as required_experience and employment_type show smaller but noticeable differences and are treated as secondary signals.

- The length of company_profile text differs substantially between fake and real jobs, with real postings containing significantly longer and more informative descriptions. This suggests that fake postings often avoid providing detailed company information even when the field is present.

- A binary feature has_urgency_language was introduced to capture urgency-related phrasing in job descriptions. While urgency language appears more frequently in fraudulent jobs (~25%) than in real jobs (~20%), the overlap indicates that this feature alone is weak and should only be used in combination with stronger signals.

- Two Named Entity Recognition–based features, org_count and gpe_count, were added to quantify how often organizations and locations are mentioned. Real job postings contain significantly more named entities on average than fake postings, indicating intentional avoidance of verifiable information by fraudulent listings. These features emerged as some of the strongest and most reliable indicators of fraud.

- A lexical diversity feature was explored to measure repetitiveness in job descriptions. However, the observed difference between fake and real jobs was small, suggesting that lexical diversity alone does not strongly distinguish fraudulent postings and is therefore deprioritized.

Final Feature Decision:
The strongest signals identified during EDA are company information availability and specificity, particularly missing company profiles and low counts of named organizations and locations. Supporting features include missing experience requirements and urgency language, while purely stylistic features such as lexical diversity are considered weak.